packages/components/src/inline-edit/hook.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * External dependencies
+ */
+import { isUndefined, negate, noop, flow } from 'lodash';
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState, useRef } from '@wordpress/element';
+
+/**
+ * @param {import("react").ChangeEvent<HTMLInputElement>} event
+ */
+const cancelEvent = ( event ) => (
+	event.preventDefault(), event.stopPropagation(), event
+);
+
+/**
+ * @param {import("react").ChangeEvent<HTMLInputElement>} event
+ */
+const getEventValue = ( { target: { value } } ) => value;
+
+/**
+ * @param {Function[]} handlers
+ */
+const mergeEvent = ( ...handlers ) =>
+	/**
+	 * @param {import("react").ChangeEvent<HTMLInputElement>} event
+	 */
+	( event ) => handlers.forEach( ( handler = noop ) => handler( event ) );
+
+/**
+ * @typedef Props
+ * @property {(value: string) => boolean} validate predicate
+ * @property {Function} onWrongInput called when a validate predicate fails
+ * @property {Function} onCommit called on enter/blur
+ * @property {string} value input value
+ */
+
+/**
+ * @param {Props} props
+ */
+export default function useInlineEdit( {
+	validate = negate( isUndefined ),
+	onWrongInput = noop,
+	onCommit = noop,
+	value: propValue,
+} ) {
+	const [ isInEditMode, setIsInEditMode ] = useState( false );
+	const [ editingValue, setEditingValue ] = useState( propValue );
+	/** @type {import('react').MutableRefObject<HTMLInputElement | undefined>} */
+	const inputRef = useRef();
+	/** @type {import('react').MutableRefObject<HTMLInputElement | undefined>} */
+	const toggleRef = useRef();
+	const isInvalid = negate( validate );
+	const changeToEditMode = () => setIsInEditMode( true );
+	const changeToToggleMode = () => setIsInEditMode( false );
+
+	useEffect( () => {
+		setEditingValue( propValue );
+		if ( isInvalid( value ) ) onWrongInput( value );
+	}, [ propValue ] );
+
+	useEffect( () => {
+		if ( isInEditMode ) {
+			inputRef.current?.focus();
+			inputRef.current?.select();
+			setEditingValue( propValue );
+		} else {
+			toggleRef.current?.focus();
+		}
+	}, [ isInEditMode ] );
+
+	/**
+	 * @param {import("react").ChangeEvent<HTMLInputElement>} event
+	 */
+	const commit = ( event ) => {
+		const { value } = event.target;
+		cancelEvent( event );
+		if ( validate( value ) ) {
+			changeToToggleMode();
+			onCommit( value );
+		} else {
+			onWrongInput( value );
+		}
+	};
+
+	/**
+	 * @param {import("react").KeyboardEvent<HTMLInputElement> & import("react").ChangeEvent<HTMLInputElement>} event
```

Comment:
Did you mean `|` instead of `&` here?

```suggestion
	 * @param {import("react").KeyboardEvent<HTMLInputElement> | import("react").ChangeEvent<HTMLInputElement>} event
```

Reply:
`&`, so I can use both `key` (from KeyEvent) and `value` from ChangeEvent